crates/rover-client/src/operations/graph/publish/runner.rs

```diff
@@ -75,36 +76,43 @@ fn build_response(
     // which very well may have changes. For this, we'll just look at the code
     // first and handle the response as if there was `None` for the diff
     let change_summary = if publish_response.code == "NO_CHANGES" {
-        build_change_summary(None)
+        ChangeSummary::none()
     } else {
-        build_change_summary(publish_response.tag.unwrap().diff_to_previous)
+        let diff = publish_response
+            .tag
+            .ok_or_else(|| RoverClientError::MalformedResponse {
+                null_field: "service.upload_schema.tag".to_string(),
+            })?
+            .diff_to_previous;
+
+        if let Some(diff) = diff {
+            build_change_summary(diff)
+        } else {
+            ChangeSummary::none()
+        }
     };
 
     Ok(GraphPublishResponse {
-        schema_hash: hash,
+        api_schema_hash: hash,
         change_summary,
     })
 }
 
-type ChangeDiff = graph_publish_mutation::GraphPublishMutationServiceUploadSchemaTagDiffToPrevious;
+type QueryChangeDiff =
+    graph_publish_mutation::GraphPublishMutationServiceUploadSchemaTagDiffToPrevious;
 
-/// builds a string-representation of the diff between two schemas
-/// e.g. ` [Fields: +2 -1 △0, Types: +4 -0 △7]` or `[No Changes]`
-fn build_change_summary(diff: Option<ChangeDiff>) -> String {
-    match diff {
-        None => "[No Changes]".to_string(),
-        Some(diff) => {
-            let changes = diff.change_summary;
-            let fields = format!(
-                "Fields: +{} -{} △ {}",
-                changes.field.additions, changes.field.removals, changes.field.edits
-            );
-            let types = format!(
-                "Types: +{} -{} △ {}",
-                changes.type_.additions, changes.type_.removals, changes.type_.edits
-            );
-            format!("[{}, {}]", fields, types)
-        }
+fn build_change_summary(diff: QueryChangeDiff) -> ChangeSummary {
+    ChangeSummary {
+        field_changes: FieldChanges {
+            additions: diff.change_summary.field.additions as u64,
+            removals: diff.change_summary.field.removals as u64,
+            edits: diff.change_summary.field.edits as u64,
+        },
+        type_changes: TypeChanges {
+            additions: diff.change_summary.type_.additions as u64,
+            removals: diff.change_summary.type_.removals as u64,
+            edits: diff.change_summary.type_.edits as u64,
+        },
```

Comment:
how do you feel about this being a method on `ChangeSummary`? For example, your `ChangeSummary` can have an instantiation `with_diff` that basically replicates this functionality you have above along with an implementation you already have for `ChangeSummary::none`. The `build_response` function can then do this:
```rust
        if let Some(diff) = diff {
            ChangeSummary::with_diff(diff)
        } else {
            ChangeSummary::none()
        }
```